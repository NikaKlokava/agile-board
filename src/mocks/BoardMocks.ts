type MockSubtaskType = {
  name: string;
};

type MockTaskType = {
  name: string;
  description: string;
  subtasks: Array<MockSubtaskType>;
};

export const MockTask: MockTaskType = {
  name: "Create something",
  description: "... it's very important ...",
  subtasks: [
    { name: "while doing 1..." },
    { name: "while doing 2..." },
    { name: "while doing 3..." },
  ],
};

export const MockBoards = [
  {
    name: "Platform Launch",
    board_column: [
      { title: "Todo", tasks: Array.from({ length: 4 }, () => MockTask) },
      { title: "Doing", tasks: Array.from({ length: 6 }, () => MockTask) },
      { title: "Done", tasks: Array.from({ length: 2 }, () => MockTask) },
    ],
  },
  {
    name: "Marketing Plan",
    board_column: [
      { title: "Todo", tasks: Array.from({ length: 7 }, () => MockTask) },
      { title: "Doing", tasks: Array.from({ length: 2 }, () => MockTask) },
      { title: "Done", tasks: Array.from({ length: 4 }, () => MockTask) },
    ],
  },
  {
    name: "Roadmap",
    board_column: [
      { title: "Now", tasks: Array.from({ length: 4 }, () => MockTask) },
      { title: "Next", tasks: Array.from({ length: 6 }, () => MockTask) },
    ],
  },
];
