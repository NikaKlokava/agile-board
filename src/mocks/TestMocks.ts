export const MockTestAddBoard = {
  name: "test name",
  columns: [{ title: "test title", uuid: "test uuid" }],
  usersEmail: ["test"],
  time: 0,
  uuid: "test",
};
