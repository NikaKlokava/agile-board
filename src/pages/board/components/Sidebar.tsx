import classes from "classnames";
import cl from "./styles/sidebar.module.css";

export const Sidebar = () => {
  return (
    <div className={cl.sidebar_wrapper}>
      <div className={cl.sidebar_title}>{"ALL BOARDS (3)"}</div>
      <div className={cl.board_title_items}>
        <div className={cl.title_item}>
          <div className={cl.icon} />
          <p className={cl.title}>Board name</p>
        </div>
        <div className={cl.title_item}>
          <div className={cl.icon} />
          <p className={cl.title}>Board name</p>
        </div>
        <div className={cl.title_item}>
          <div className={cl.icon} />
          <p className={cl.title}>Board name</p>
        </div>
        <div className={cl.title_item}>
          <div className={cl.icon_new_board} />
          <p className={classes(cl.title, cl.new_board)}>New board</p>
        </div>
      </div>
      <div className={cl.hide_sidebar}>
        <div className={cl.hide_icon} />
        <div className={cl.hide_text}>Hide sidebar</div>
      </div>
    </div>
  );
};
