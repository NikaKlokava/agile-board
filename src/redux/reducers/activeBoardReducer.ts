import {
  DELETE_ACTIVE_BOARD_ACTION,
  SELECT_BOARD_ACTION,
} from "../actions/actions";

const initialState = { uuid: "", name: "", columns: [] };

export const activeBoardReducer = (
  state = initialState,
  action: SelectBoardActionType
) => {
  switch (action.type) {
    case SELECT_BOARD_ACTION:
      return {
        ...state,
        ...action.payload.board,
      };
    case DELETE_ACTIVE_BOARD_ACTION:
      return { ...state, ...initialState };
    default:
      return state;
  }
};
