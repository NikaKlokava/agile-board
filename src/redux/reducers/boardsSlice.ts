import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import { v4 as uuidv4 } from "uuid";

const initialState: BoardsType | never[] = {
  boards: [],
};

export const boardsSlice = createSlice({
  name: "boards",
  initialState,
  reducers: {
    addBoard: (
      state: BoardsType,
      action: PayloadAction<{ name: string; columns: { title: string }[] }>
    ) => {
      const newBoard = {
        ...action.payload,
        uuid: uuidv4(),
        columns: action.payload.columns.map((column) => {
          return {
            ...column,
            uuid: uuidv4(),
          };
        }),
      };
      state.boards = [...state.boards, newBoard];
    },
    addNewColumn: (
      state: BoardsType,
      action: PayloadAction<{
        name: string;
        columns: { title: string; uuid?: string }[];
        uuid: string | undefined;
      }>
    ) => {
      const newColumns = action.payload.columns.map((column) => {
        const newColumn = {
          ...column,
          uuid: uuidv4(),
        };
        if (!column.uuid) return newColumn;
        return column;
      });
      state.boards = state.boards.map((board: BoardType) => {
        const updatedBoard = {
          ...board,
          name: action.payload.name,
          columns: newColumns,
        };
        if (board.uuid === action.payload.uuid) return updatedBoard;
        return board;
      });
    },
    deleteBoard: (
      state: BoardsType,
      action: PayloadAction<{ uuid: string }>
    ) => {
      state.boards = state.boards.filter(
        (board: BoardType) => board.uuid !== action.payload.uuid
      );
    },
  },
});

export const { addBoard, addNewColumn, deleteBoard } = boardsSlice.actions;

export default boardsSlice.reducer;
