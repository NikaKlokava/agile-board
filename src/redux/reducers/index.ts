import { combineReducers } from "redux";
import { activeBoardReducer } from "./activeBoardReducer";
import { boardsReducer } from "./boardsReducer";
import { tasksReducer } from "./tasksReducer";

export const rootReducer = combineReducers({
  boards: boardsReducer,
  tasks: tasksReducer,
  activeBoard: activeBoardReducer,
});
