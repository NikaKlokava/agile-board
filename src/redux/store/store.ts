import { legacy_createStore } from "redux";
import { boardsReducer } from "../reducers/boardReducer";

const store = legacy_createStore(boardsReducer);

export default store;

// export type RootState = ReturnType<typeof store.getState>

// export interface RootState {
//     user: UserState;
//     loader: LoaderState;
//   }
// export type AppDispatch = typeof store.dispatch
