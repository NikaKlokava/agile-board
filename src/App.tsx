import { Provider } from "react-redux";
import { Route, Routes } from "react-router";
import "./index.css";
import { BoardPage } from "./pages/board";
import { ErrorPage } from "./pages/error";
import store from "./redux/store/store";

function App() {
  return (
    <Provider store={store}>
      <Routes>
        <Route path="/agile-board" element={<BoardPage />} />
        <Route path="*" element={<ErrorPage />} />
      </Routes>
    </Provider>
  );
}

export default App;
