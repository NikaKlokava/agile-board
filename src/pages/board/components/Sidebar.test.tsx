import { render, screen, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Sidebar } from "./Sidebar";
import { Provider } from "react-redux";
import store from "../../../redux/store/store";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import App from "../../../App";
import { addBoard, changeStatus } from "../../../redux/reducers/boardsSlice";
import { MockBoard } from "../../../mocks/BoardMocks";
jest.mock("../../../shared/hooks/useAuthorization", () => ({
  useAuthorization: () => ({ isUserExist: true }),
}));

beforeEach(() => jest.clearAllMocks());

describe("Test the Sidebar component", () => {
  test("The Sidebar renders correctly", () => {
    const sidebarSnap = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <Sidebar />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(sidebarSnap).toMatchSnapshot();
  });
});

describe("Test the New Board element", () => {
  test("The NewBoardModal should be visible on click", async () => {
    render(<App />);
    act(() => {
      store.dispatch(changeStatus({ isLoading: false }));
      store.dispatch(addBoard(MockBoard));
    });

    await waitFor(() => {
      const newBoardEl = screen.getByTestId("new-board");
      newBoardEl.click();
      const newBoardModal = screen.queryByTestId("new-board-modal");

      expect(newBoardModal).toBeInTheDocument();
    });
  });
});
