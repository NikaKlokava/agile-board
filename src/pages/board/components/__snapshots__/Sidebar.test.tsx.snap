// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Sidebar component The Sidebar renders correctly 1`] = `
<div
  className="sidebar_wrapper"
>
  <div
    className="sidebar_title"
  >
    ALL BOARDS (0)
  </div>
  <div
    className="board_title_items"
  >
    <div
      className="title_item"
      data-testid="new-board"
      onClick={[Function]}
    >
      <div
        className="icon_new_board"
      />
      <p
        className="title new_board"
      >
        New board
      </p>
    </div>
  </div>
  <div
    className="hide_sidebar"
  >
    <div
      className="hide_icon"
    />
    <div
      className="hide_text"
    >
      Hide sidebar
    </div>
  </div>
</div>
`;
